File: shared-utils/pow-utils/manual-tests/debug_messages.py
#!/usr/bin/env python3
# SPDX-License-Identifier: Apache-2.0
"""
Debug script to print out exact byte sequences for comparison
"""

import torch
import hashlib

def _tok_le_bytes(tok_i64: torch.Tensor) -> torch.ByteTensor:
    """View an (B, L) int64 tensor as little-endian bytes."""
    t = tok_i64.contiguous()
    return t.view(torch.uint8).view(t.size(0), -1)

def _u32le(x: torch.Tensor) -> torch.ByteTensor:
    """Convert a 32-bit tensor to little-endian bytes."""
    x32 = x.reshape(-1).to(torch.uint32).contiguous()
    x8 = x32.view(torch.uint8)
    return x8.view(-1, 4)

def _str_bytes(s: str, batch_size: int, device='cpu') -> torch.ByteTensor:
    """Convert string to bytes tensor."""
    b = s.encode('utf-8')
    arr = torch.tensor(list(b), dtype=torch.uint8, device=device)
    return arr.unsqueeze(0).expand(batch_size, -1)

def hex_to_bytes_tensor(hex_str: str, device='cpu') -> torch.ByteTensor:
    """Convert hex string to ByteTensor."""
    bytes_data = bytes.fromhex(hex_str)
    return torch.tensor(list(bytes_data), dtype=torch.uint8, device=device)

# Test exact same values as C++ test
h_b = hex_to_bytes_tensor("0" * 63 + "1")
v = hex_to_bytes_tensor("0" * 63 + "2")
tick = 100
step = 42

# Context tokens
context_tokens = torch.tensor([[1234, 5678]], dtype=torch.int64)
ctx_bytes = _tok_le_bytes(context_tokens)

print("=== Python Debug Output ===")
print(f"h_b bytes: {h_b.numpy().tobytes().hex()}")
print(f"v bytes: {v.numpy().tobytes().hex()}")
print(f"Context tokens: {context_tokens[0].tolist()}")
print(f"Context bytes ({ctx_bytes.shape[1]} bytes): {ctx_bytes[0].numpy().tobytes().hex()}")

# Build message components
j4 = _u32le(torch.tensor([step], dtype=torch.int32).view(-1, 1))
print(f"Step (j4) bytes: {j4[0].numpy().tobytes().hex()}")

# IMPORTANT: T8 should be 8 bytes, but we're converting from a 32-bit tick value
T8_32 = _u32le(torch.tensor([tick], dtype=torch.uint32))
T8 = torch.zeros(1, 8, dtype=torch.uint8)
T8[0, :4] = T8_32[0]  # First 4 bytes are the tick in little-endian
# Last 4 bytes remain zeros
print(f"Tick (T8) bytes (8 bytes total): {T8[0].numpy().tobytes().hex()}")

precision = _str_bytes("fp16", batch_size=1)
print(f"Precision bytes: {precision[0].numpy().tobytes().hex()}")

# Build complete message
# Order: header_prefix (32), v (32), T8 (8), j4 (4), ctx_bytes (16), precision (4)
msg_parts = {
    "h_b (32)": h_b.numpy().tobytes().hex(),
    "v (32)": v.numpy().tobytes().hex(), 
    "T8 (8)": T8[0].numpy().tobytes().hex(),
    "j4 (4)": j4[0].numpy().tobytes().hex(),
    "ctx_bytes (16)": ctx_bytes[0].numpy().tobytes().hex(),
    "precision (4)": precision[0].numpy().tobytes().hex()
}

print("\nMessage components in order:")
total_bytes = 0
for name, hex_val in msg_parts.items():
    byte_count = len(hex_val) // 2
    total_bytes += byte_count
    print(f"  {name}: {hex_val} ({byte_count} bytes)")

# Build message
B = ctx_bytes.size(0)
msg = torch.cat([
    h_b.view(1, -1).expand(B, -1),      # 32 bytes
    v.view(1, -1).expand(B, -1),        # 32 bytes  
    T8,                                  # 8 bytes
    j4,                                  # 4 bytes
    ctx_bytes,                           # 16 bytes (2 tokens * 8 bytes each)
    precision,                           # 4 bytes
], dim=1)

print(f"\nTotal message length: {msg.shape[1]} bytes (expected: {total_bytes})")
print(f"Complete message hex:\n{msg[0].numpy().tobytes().hex()}")

# Compute hash
msg_bytes = msg[0].cpu().numpy().tobytes()
digest = hashlib.sha256(msg_bytes).digest()
print(f"\nSHA256 digest: {digest.hex()}")

# Compute U value
digest_tensor = torch.tensor(list(digest), dtype=torch.uint8)
b0 = digest_tensor[0].to(torch.float32)
b1 = digest_tensor[1].to(torch.float32) 
b2 = digest_tensor[2].to(torch.float32)
b3 = digest_tensor[3].to(torch.float32)
u = (b0 + b1 * 256 + b2 * 65536 + b3 * 16777216) / 4294967296.0
print(f"U value: {u:.10f}")

# Test CDF sampling
cdf = torch.tensor([0.1, 0.3, 0.6, 0.8, 0.9, 0.95, 0.99, 1.0])
token_id = torch.searchsorted(cdf, torch.tensor(u))
print(f"Token ID: {token_id.item()}")

File: shared-utils/pow-utils/debug_messages.cpp
// SPDX-License-Identifier: Apache-2.0
// debug_messages.cpp - Debug message construction
#include "pow_utils.h"
#include <iostream>
#include <iomanip>
#include <openssl/sha.h>

int main() {
    std::cout << "=== C++ Debug Output ===" << std::endl;
    
    // Test exact same values as Python
    auto h_b = hex_to_bytes("0000000000000000000000000000000000000000000000000000000000000001");
    auto v = hex_to_bytes("0000000000000000000000000000000000000000000000000000000000000002");
    int tick = 100;
    int32_t step = 42;
    
    std::cout << "h_b bytes: " << bytes_to_hex(h_b) << std::endl;
    std::cout << "v bytes: " << bytes_to_hex(v) << std::endl;
    
    // Context tokens
    std::vector<int64_t> context = {1234, 5678};
    std::cout << "Context tokens: [" << context[0] << ", " << context[1] << "]" << std::endl;
    
    auto ctx_bytes = tok_le_bytes(context);
    std::cout << "Context bytes (" << ctx_bytes.size() << " bytes): " << bytes_to_hex(ctx_bytes) << std::endl;
    
    // Step bytes
    auto j4 = u32le(step);
    std::cout << "Step (j4) bytes: " << bytes_to_hex(j4) << std::endl;
    
    // Tick bytes - IMPORTANT: Should be 8 bytes total!
    auto T8_32 = u32le(tick);
    std::vector<uint8_t> T8(8, 0);  // 8 bytes, initialized to 0
    std::copy(T8_32.begin(), T8_32.end(), T8.begin());  // Copy first 4 bytes
    std::cout << "Tick (T8) bytes (8 bytes total): " << bytes_to_hex(T8) << std::endl;
    
    // Precision bytes
    auto precision = str_bytes("fp16");
    std::cout << "Precision bytes: " << bytes_to_hex(precision) << std::endl;
    
    // Build message components
    std::cout << "\nMessage components in order:" << std::endl;
    std::cout << "  h_b (32): " << bytes_to_hex(h_b) << " (" << h_b.size() << " bytes)" << std::endl;
    std::cout << "  v (32): " << bytes_to_hex(v) << " (" << v.size() << " bytes)" << std::endl;
    std::cout << "  T8 (8): " << bytes_to_hex(T8) << " (" << T8.size() << " bytes)" << std::endl;
    std::cout << "  j4 (4): " << bytes_to_hex(j4) << " (" << j4.size() << " bytes)" << std::endl;
    std::cout << "  ctx_bytes (16): " << bytes_to_hex(ctx_bytes) << " (" << ctx_bytes.size() << " bytes)" << std::endl;
    std::cout << "  precision (4): " << bytes_to_hex(precision) << " (" << precision.size() << " bytes)" << std::endl;
    
    // Build complete message
    auto msg = build_msg(h_b, v, T8, j4, ctx_bytes, precision);
    
    std::cout << "\nTotal message length: " << msg.size() << " bytes" << std::endl;
    std::cout << "Complete message hex:\n" << bytes_to_hex(msg) << std::endl;
    
    // Compute hash
    unsigned char hash[SHA256_DIGEST_LENGTH];
    SHA256_CTX sha256;
    SHA256_Init(&sha256);
    SHA256_Update(&sha256, msg.data(), msg.size());
    SHA256_Final(hash, &sha256);
    
    std::vector<uint8_t> digest(hash, hash + SHA256_DIGEST_LENGTH);
    std::cout << "\nSHA256 digest: " << bytes_to_hex(digest) << std::endl;
    
    // Compute U value
    float u = digest_to_u(digest);
    std::cout << "U value: " << std::fixed << std::setprecision(10) << u << std::endl;
    
    // Test CDF sampling
    std::vector<float> cdf = {0.1f, 0.3f, 0.6f, 0.8f, 0.9f, 0.95f, 0.99f, 1.0f};
    auto it = std::lower_bound(cdf.begin(), cdf.end(), u);
    int64_t token_id = std::distance(cdf.begin(), it);
    std::cout << "Token ID: " << token_id << std::endl;
    
    return 0;
}